src/main/java/org/aksw/iguana/cc/worker/impl/SPARQLProtocolWorker.java

```diff
@@ -0,0 +1,412 @@
+            POST_UPDATE("post update");
+
+            private final String value;
+
+            @JsonCreator
+            RequestType(String value) {
+                this.value = Objects.requireNonNullElse(value, "get query");
+            }
+
+            @JsonValue
+            public String value() {
+                return value;
+            }
+        }
+
+        private final RequestType requestType;
+
+        public RequestFactory(RequestType requestType) {
+            this.requestType = requestType;
+        }
+
+        private static String urlEncode(List<String[]> parameters) {
+            return parameters.stream()
+                    .map(e -> e[0] + "=" + URLEncoder.encode(e[1], StandardCharsets.UTF_8))
+                    .collect(Collectors.joining("&"));
+        }
+
+        public HttpRequest buildHttpRequest(InputStream queryStream,
+                                            Duration timeout,
+                                            ConnectionConfig connection,
+                                            String requestHeader) throws URISyntaxException, IOException {
+            HttpRequest.Builder request = HttpRequest.newBuilder().timeout(timeout);
+
+            class CustomStreamSupplier {
+                boolean used = false; // assume, that the stream will only be used again, if the first request failed, because of the client
+                public Supplier<InputStream> getStreamSupplier() {
+                    if (!used) {
+                        used = true;
+                        return () -> queryStream;
+                    }
+                    else
+                        return () -> null;
+                }
+            }
+
+            final var streamSupplier = new CustomStreamSupplier();
+
+            if (requestHeader != null)
+                request.header("Accept", requestHeader);
+            if (connection.authentication().user() != null)
+                request.header("Authorization",
+                               HttpWorker.basicAuth(connection.authentication().user(),
+                                                    Optional.ofNullable(connection.authentication().password()).orElse("")));
+            switch (this.requestType) {
+                case GET_QUERY -> {
+                    request.uri(new URIBuilder(connection.endpoint())
+                                    .setParameter("query",
+                                            new String(queryStream.readAllBytes(), StandardCharsets.UTF_8))
+                                    .build())
+                            .GET();
+                }
+                case POST_URL_ENC_QUERY -> {
+                    request.uri(connection.endpoint())
+                            .header("Content-Type", "application/x-www-form-urlencoded")
+                            .POST(HttpRequest.BodyPublishers.ofString(
+                                    urlEncode(Collections.singletonList(
```

Comment:
(1): create an issue to either url-encode on the fly or pre-encode the query string and supply it encoded as stream